### PR TITLE
Fix exported typescript typings for version 2.0

### PR DESCRIPTION
**Pull request by HarelM**
It seems that MapLibre is working as expected when it comes to javascript but the exported typescript typings have changed significantly due to the migration to typescript and they need to be fixed.
As a reference, I'm looking at the original typescript typings that existed in this project in version 1.x:
https://github.com/maplibre/maplibre-gl-js/blob/cb3a4200bfbc113969e9dcbedae2aa7655f9aa98/src/index.d.ts
Which has a lot more types and exported stuff compared to the file generated in types folder after running build-tsc.

PR Summary to address (#610) last issues:
1. Improve typing - especially the map movement types, and events.
2. Remove types folder as it is no longer needed
3. Introduce `dts-build-generator` to bundle all the typescript definitions files to a single one
4. Rename copy-glsl.js and added the functionality of the bundling and some fixed after bundling
5. Added a `package.json` file in the dist folder to signal external builder (such as webpack) to treat the output as commonjs (#370)
6. Renames style spec's `es.js` to `.mjs` to allow bundlers to know which type of module this code is expecting to be used.

### CODE REVIEW

**Comment by HarelM** [comment]
I've posted a request for help here:
https://github.com/microsoft/TypeScript/issues/10907

**Comment by HarelM** [comment]
Another thought about this is the following article about merging the definitions file into a single file:
https://medium.com/@martin_hotell/typescript-library-tips-rollup-your-types-995153cc81c7
An in general a post build step to fix whatever we need.
I still need to think about this...

**Comment by wipfli** [comment]
Thanks for the testing and for discovering this problem Harel!

**Comment by wipfli** [comment]
How do we actually create `index.d.ts`? It seems we are using neither `rollup-plugin-dts` nor `api-extractor` as suggested in the medium article you shared.

**Comment by HarelM** [comment]
I must admit I'm leaning towards migrating to named exports instead of keeping this "old" method of exporting a default object.
In most libraries I saw the direction was to use named exports instead of default (I think this is to allow tree shaking in general).
I don't know if it will help this issue or not, but if we are already releasing 2.x and breaking things, I tend towards going in this direction and migrating to named exports.
@wipfli @astridx @xabbu42 What do you think?

**Comment by HarelM** [comment]
`index.d.ts` is created by the `tsc` command, when you configure `declaration` in the `tsconfig` file.
I also saw that this broke due to the changes made in the benchmark effort as the typings moved a folder down (much like the js code itself). I need to fix this as well...

**Comment by wipfli** [comment]
I guess we have to change this is what you are saying: 

https://github.com/maplibre/maplibre-gl-js/blob/50cb89270ee0ff1459d49dc556b921589af3828a/tsconfig.json#L25

**Comment by HarelM** [comment]
Nope, I mean this line is now incorrect:
https://github.com/maplibre/maplibre-gl-js/blob/50cb89270ee0ff1459d49dc556b921589af3828a/package.json#L15
It probably needs to be 
`types/src/index.d.ts`, 
I think pre.5 was released before the bench fixes and thus it was still valid, right?

**Comment by xabbu42** [comment]
This is too far outside my expertise, so I defer to your judgement on that one.

**Comment by wipfli** [comment]
We released pre.5 on Sep 25, 2021 and merged the benchmark stuff on Oct 6, 2021.

**Comment by wipfli** [comment]
So, yes, pre.5 is still valid @HarelM.

**Comment by wipfli** [comment]
I always like it when code is explicit, and named exports are going in that direction. Is `Map` the only object that gets exported by this library by default and would have to be replaced with a named export?

**Comment by wipfli** [comment]
I am confused by the fact that the `.d.ts` file you show in the typescript issue linked above is only a few lines long while the old one you also linked above has like 2k lines. How comes the new one is so much shorter?

**Comment by HarelM** [comment]
>Is Map the only object that gets exported by this library by default and would have to be replaced with a named export?

No, the object that gets exported is an object with multiple field, `Map` is just one of them.

>I am confused by the fact that the .d.ts file you show in the typescript issue linked above

This is the entry point file which reference all the d.ts files in that folder, having said that, there are many many missing exported types and methods that existed in the original typescript definition file that we'll need to make sure are exported now properly before this issue is resolved.

Another possible solution to this issue is simply bring back the definitions file from version 1.x and avoid all this declaration generation stuff, but it's a workaround and not a maintainable solution from my point of view.

Named exports is the right direction, but I'm not sure I'll be able to migrate this library to use it:
Currently when exporting the "main" object there are some getters and setters that are exported too, it is some kind of global configuration singleton object, I'm not sure what will break if I export this global configuration object as a named export and what should be the right way to export it.
It seems that this library is written in a way that will be very hard to change in order to benefit from newer technologies and standards... :-(

**Comment by wipfli** [comment]
Can you give me an example for a configuration singleton in the old `index.d.ts` file?

**Comment by HarelM** [comment]
`addProtocol`, `workerCount`, `maxParallelImageRequests` etc. Look in the `src/index.ts` file and you'll see that below the exported classes there are a lot of "global" functions.

**Comment by wipfli** [comment]
OK, thanks. The maplibre namespace global functions are:

```js
    export function supported(options?: { failIfMajorPerformanceCaveat?: boolean }): boolean;

    export function clearStorage(callback?: (err?: Error) => void): void;
   
    export function addProtocol(customProtocol: string, loadFn: (requestParameters: RequestParameters, callback: ResponseCallback<any>) => Cancelable): void;

    export function removeProtocol(customProtocol: string): void;

    export function setRTLTextPlugin(pluginURL: string, callback: (error: Error) => void, deferred?: boolean): void;

    export function getRTLTextPluginStatus(): PluginStatus;

    export function prewarm(): void;

    export function clearPrewarmedResources(): void;
```

And the namespace global variables are:

```js
    let accessToken: string;
    let version: string;
    let baseApiUrl: string;
    let workerCount: number;
    let maxParallelImageRequests: number;
```

Are these variables the problem? If so, should we make them private and replace with setters and getters?

**Comment by HarelM** [comment]
Are you looking in the recent typescript code or the "old" declaration file?
It's not about getters and setters but rather how to export this "the right way" when it comes to named exports...

**Comment by wipfli** [comment]
I was looking into the old file.

**Comment by wipfli** [comment]
In yesterday's meeting we discussed the option of migrating the library to named exports. This requires code changes in the library, and it means that users have to change the way they import MapLibre GL JS. The benefit of named exports is that users will only include the parts of the library which they need, so that their overall file size gets smaller.

**Comment by HarelM** [comment]
I wanted to go a step back and see if I can somehow make a post processing of some kind to the exported `index.d.ts` file in order to make it "correct" - i.e. similar to how it was in version 1.x with the `namespace` and `export =`.
Since you can't re-export with "*" into the namespace (or at least I didn't find a way) it means that in order to make this post processing work there will be a need to re-export all the types of the project into the namespace, or use the old file and risk this getting out of sync. Both options seems bad to me...
I can't help but blame umd format for this, as it seems that typescript is not favoring it in any way...

**Comment by HarelM** [comment]
My current efforts are focused in making the types better so that the following tool will allow generating a umd module type definitions.
See the following issue:
https://github.com/timocov/dts-bundle-generator/issues/175
#641 and #639 are related to this effort, I don't know if it will succeed but making the types better is a nice side benefit either way :-)

**Comment by HarelM** [comment]
I did a hack to export all the classes in the generated typings file using the following command line (macOS):
`npx dts-bundle-generator --umd-module-name=maplibregl -o mlgl-tmp.d.ts ./src/index.ts ; awk '{gsub(/declare class/,"export declare class")}1' mlgl-tmp.d.ts > index.d.ts`
This creates a typings file which has most of what's needed.
Still to go:
1. Canvas source: for some reason it's not part of the generated code, I'm not sure why...
2. Some of the events that are defined in `events.ts` are not exported.

I feel like this is the last stretch, I managed to compile a lot of the ngx-maplibre-gl code using this definition file and I see a very good correlation between the types that are defined in this code and the exported types - i.e. if the types here are bad we get a bad definition file which leads me to fix some of the bad typings in this codebase... :-)

**Comment by HarelM** [comment]
P.S. named exports will probably not be part of this effort as the build is complicated and changing to named exports basically breaks everything...
The following is the history of how the build got to where it is today I believe:
https://github.com/rollup/rollup/issues/1029

**Comment by wipfli** [comment]
Beautiful, go Harel :dolphin:

**Comment by HarelM** [comment]
Last commit adds some event typings from the original `d.ts` file and seems to improve the generated typings.
I've also added some overloads to the `on(...)` in the `map.ts` which causes the relevant event typings to be in the generated file.
I've managed to compile the `ngx-maplibre-gl` project and replace all the relevant changes related to the typings but it still doesn't fully compile, I hope I'll be able to address this soon... I still think the typings are not equivalent to the original and this creates some issues, I'll see if I can manually fix this to see what need to be done...

File: CHANGELOG.md [hunk 1]
```diff
@@ -19,7 +17,22 @@
 - ** Breaking Change ** stopped supporting IE (internet explorer)
 - ** Breaking Change ** removed all code related to `accessToken` and mapbox specific urls, including telemetry etc.  Please do not use mapbox servers with this library.
 - ** Breaking Change ** removed `baseApiUrl` as it was used only for mapbox related urls
-- Added redraw function to map (#206)
+- ** Breaking Change ** typescript tyings have changed:
```
Comment (wipfli) [inline CHANGELOG.md:20]: typo in `typings`

Comment (HarelM) [reply CHANGELOG.md]: Fixed. Thanks!

**Comment by HarelM** [comment]
All tests are passing :-) I've highlighted the changes I made in the first comment of this PR.
Feel free to review.
Once reviewed and merged we can create a `pre.6` version to test it. I Hope the fixes here are the last fixes needed in order to be able to release version 2.0. :-)

File: src/index.ts [hunk 2]
```diff
@@ -1,3 +1,5 @@
+/// <reference path="./types/glsl.d.ts" />
+/// <reference path="./types/non-typed-modules.d.ts" />
```
Comment (wipfli) [inline src/index.ts:2]: What is this?

File: src/ui/camera.ts [hunk 3]
```diff
@@ -14,6 +14,8 @@ import type {LngLatBoundsLike} from '../geo/lng_lat_bounds';
 import type {TaskID} from '../util/task_queue';
 import type {PaddingOptions} from '../geo/edge_insets';
 
+export type RequireAtLeastOne<T> = { [K in keyof T]-?: Required<Pick<T, K>> & Partial<Pick<T, Exclude<keyof T, K>>>; }[keyof T];
```
Comment (wipfli) [inline src/ui/camera.ts:17]: Looks cool. Can you comment what this does?

**Review by wipfli — APPROVED** [review]
Looks roughly correct to me. Thanks Harel!

**Comment by wipfli** [comment]
Is there a best practice for the exported types? I am a bit surprised that they are not covered by unit tests.

File: src/index.ts (hunk 2 above)
Comment (HarelM) [reply src/index.ts:2]: This is a way to reference files in typescript. The tsc does this automatically but it is needed for the bundle-generator tool to work correctly

File: src/ui/camera.ts (hunk 3 above)
Comment (HarelM) [reply src/ui/camera.ts:17]: A generic type to allow defining a type that has at least one required field (i.e. one field that is mandatory).
I don't know how to explain this better, can you suggest a comment that will make it clear for you?

**Comment by HarelM** [comment]
I think that once we convert all out unit test to typescript we may be able to cover this in the test-build suite to make sure some types are exported as expected. I hope this is what you mean...

File: src/ui/camera.ts (hunk 3 above)
Comment (wipfli) [reply src/ui/camera.ts:17]: In that case you gave it actually the right name and we can leave it as is I think.

**Comment by wipfli** [comment]
To create a release, got to the GitHub Actions tab and trigger the "Release" workflow manually.

**Comment by HarelM** [comment]
Way ahead of you :-) version pre.6 was just released.
Seems like the typings and module issue is resolved from a simple test I just made.
I'll see if I can pre-release ngx-maplibre-gl too and use it and pre.6 in my main project.

**Pull request merged**